chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/state/JoinedRoomsStateDiffer.kt

```diff
@@ -0,0 +1,86 @@
+package com.pusher.chatkit.rooms.state
+
+import com.pusher.chatkit.state.Action
+import com.pusher.chatkit.state.JoinedRoom
+import com.pusher.chatkit.state.LeftRoom
+import com.pusher.chatkit.state.RoomUpdated
+import com.pusher.chatkit.state.State
+import org.reduxkotlin.GetState
+
+internal class JoinedRoomsStateDiffer(val stateGetter: GetState<State>) {
+
+    internal fun stateExists() = stateGetter().joinedRoomsState != null
+
+    internal fun toActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> {
+        return joinedRoomActions(newRooms, newUnreadCounts) +
+            roomUpdatedActions(newRooms) +
+            leftRoomActions(newRooms)
+    }
+
+    private val currentRooms get() = stateGetter().joinedRoomsState!!.rooms
+
+    private fun joinedRoomActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> {
+        val actions = mutableListOf<Action>()
+
+        val addedRoomKeys =
+            newRooms.map { it.id }.toSet() - currentRooms.keys.toSet()
+
+        for (addedKey in addedRoomKeys) {
+            val room = newRooms.find { it.id == addedKey }
+            val unreadCount = newUnreadCounts[addedKey]
+
+            if (room != null &&
```

Comment:
This room should never be null. Such `if` could hide a programming mistake so it's better to use `!!` as explicit expectation. Generally this code could be much simpler, shorter and cleaner as processing stream (functional, not imperative programming style). V1 is nicely written in that way and we should keep those standards up.

Reply:
👍